### PR TITLE
Prevent multiple return statements in a loop when replacing expressions

### LINKED ISSUE #5022: Do expressions transform incorrectly for `if` inside `while` statements
### Input Code

```js
let a = do {
  while (1) {
    if (a) {
      a
    } else {
      b
    }
  }
};

```

### Babel Configuration (.babelrc, package.json, cli command)

```js
{
  "presets": [
    "es2015",
    "stage-0"
  ]
}
```

### Expected Behaviour
```js
let a = function () {
  var _ret;

  while (1) {
    if (a) {
      _ret = a;
    } else {
      _ret = b;
    }
  }
  return _ret;
}();
```

### Current Behaviour
```js
let a = function () {
  var _ret, _ret2;

  while (1) {
    if (a) {
      _ret = a;
    } else {
      _ret2 = b;
    }
  }
  return _ret;
  return _ret2;
}();
```

### Your Environment

| software         | version
| ---------------- | -------
| Babel            |  6.19.0
| node             | 6.9.2
| npm              | 3
| Operating System | Mac OS X 10.11

### PR DESCRIPTION
**Pull request by existentialism**
| Q                        | A 
| ------------------------ | ---
| Patch: Bug Fix?          |  Y
| Major: Breaking Change?  | N
| Minor: New Feature?      | N
| Deprecations?            | N
| Spec Compliancy?         | N
| Tests Added/Pass?        | Y/Y
| Fixed Tickets            | Fixes #5022 
| License                  | MIT
| Doc PR                   | 
| Dependency Changes       | 

Could definitely use some guidance on this, as I'm still digging through a lot of this :P

### CODE REVIEW

**Comment by xtuc** [comment]
Nice job! :+1:

File: packages/babel-traverse/src/path/replacement.js [hunk 1]
```diff
@@ -212,10 +212,14 @@ export function replaceExpressionWithStatements(nodes: Array<Object>) {
...
       let loop = path.findParent((path) => path.isLoop());
       if (loop) {
-        let callee = this.get("callee");
-
-        let uid = callee.scope.generateDeclaredUidIdentifier("ret");
-        callee.get("body").pushContainer("body", t.returnStatement(uid));
+        let uid = loop.getData("expressionReplacementReturnUid");
+
+        if (!uid) {
+          let callee = this.get("callee");
```
Comment (Kovensky) [inline packages/babel-traverse/src/path/replacement.js]: `const`

File: packages/babel-plugin-transform-do-expressions/test/fixtures/do-expressions/while-if/actual.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+let a = do {
+  while (p = p.parentPath) {
```
Comment (Kovensky) [inline packages/babel-plugin-transform-do-expressions/test/fixtures/do-expressions/while-if/actual.js]: I'd feel better if there was a `let p` before the `let a` 😃

File: packages/babel-traverse/src/path/replacement.js [hunk 3]
```diff
@@ -212,10 +212,14 @@ export function replaceExpressionWithStatements(nodes: Array<Object>) {
...
-        let uid = callee.scope.generateDeclaredUidIdentifier("ret");
-        callee.get("body").pushContainer("body", t.returnStatement(uid));
+        let uid = loop.getData("expressionReplacementReturnUid");
+
+        if (!uid) {
+          const callee = this.get("callee");
+          uid = callee.scope.generateDeclaredUidIdentifier("ret");
+          callee.get("body").pushContainer("body", t.returnStatement(uid));
+          loop.setData("expressionReplacementReturnUid", uid);
+        }
```
Comment (loganfsmyth) [inline packages/babel-traverse/src/path/replacement.js]: We should probably do

```
} else {
  uid = t.identifier(uid.name);
}
```
or else we'll be inserting the same identifier node in multiple places in the AST.

**Comment by loganfsmyth** [comment]
One small comment, and this has a conflict, but otherwise looks good.

File: packages/babel-traverse/src/path/replacement.js (hunk 3 above)
Comment (existentialism) [inline packages/babel-traverse/src/path/replacement.js]: 👍

**Pull request merged**